ui/insight/src/chart.ts

```diff
@@ -235,20 +252,31 @@ function makeChart(el, data) {
   Highcharts.chart(el, chartConf);
 }
 
-function empty(txt) {
-  return m('div.chart.empty', [m('i[data-icon=]'), txt]);
+function empty(txt: string) {
+  return h('div.chart.empty', [
+    h('i', {
+      attrs: { 'data-icon': '' },
+    }),
+    txt,
+  ]);
+}
+
+function chartHook(vnode: VNode, ctrl: Ctrl) {
+  const el = vnode.elm as HTMLElement;
+  if (ctrl.vm.loading || !ctrl.vm.answer) {
+    $(el).html(lichess.spinnerHtml);
+  } else {
+    makeChart(el, ctrl.vm.answer);
+  }
 }
 
-module.exports = function (ctrl) {
+export default function (ctrl: Ctrl) {
   if (!ctrl.validCombinationCurrent()) return empty('Invalid dimension/metric combination');
-  if (!ctrl.vm.answer.series.length) return empty('No data. Try widening or clearing the filters.');
-  return [
-    m('div.chart', {
-      config: function (el) {
-        if (ctrl.vm.loading) return;
-        makeChart(el, ctrl.vm.answer);
-      },
-    }),
-    ctrl.vm.loading ? m.trust(lichess.spinnerHtml) : null,
```

Comment:
What happened to the loading spinner? You can use this as reference: <https://github.com/ornicar/lila/blob/master/ui/analyse/src/util.ts#L81>. Probably should get a module in `ui/common`.

Reply:
Good question; it's added in a hook over here https://github.com/370417/lila/blob/39c51a5ae64c59413e1699f2a1031361da80b00b/ui/insight/src/chart.ts#L267